Implement JWT Authentication with Refresh Token Flow and Secure Logout

AmmarAbdelateif0-0 opened the pull request:
<img width="1088" height="581" alt="image" src="https://github.com/user-attachments/assets/25171021-1fa7-46ac-85fb-8abc456cce63" />

Refactor auth, add refresh tokens & patient CRUD APIs

- Implement refresh token support and robust JWT config
- Split registration by role; add role-specific DTOs
- Update EF Core models, migrations for new relationships
- Improve error handling, logging, and validation
- Update dependencies and appsettings; remove Keycloak JSON
-  Added Patient, Doctor, and Caregiver entities extending ApplicationUser
- Created role-specific registration endpoints (`/register/patient`, `/register/doctor`, `/register/caregiver`)
- Implemented dedicated DTOs for each user type with appropriate validations
- Added AuthService with role-specific business logic
- Configured Entity Framework relationships between user types

The Workflow : 
<img width="2110" height="706" alt="image" src="https://github.com/user-attachments/assets/fecf90cb-6888-49a9-a0a0-88da20770576" />

AmmarAbdelateif0-0 committed:
Integrate ASP.NET Core Identity for user management

Refactored the codebase to use ASP.NET Core Identity, replacing the custom User class and ApplicationDbContext with ApplicationUser, ApplicationRole, and BreastCancerDB (IdentityDbContext). Updated all repositories and services to use the new context. Added and updated EF Core migrations to create Identity tables and seed default roles. Patient, Doctor, and Caregiver now inherit from ApplicationUser. Removed the old User class and all references. Updated Program.cs to configure Identity services and appsettings.json for local development. The solution now supports Identity-based authentication and role management.

AmmarAbdelateif0-0 committed:
Replace Keycloak with ASP.NET Identity & JWT Auth

Refactored authentication to use ASP.NET Core Identity and JWT tokens instead of Keycloak. Added AccountController, DTOs, and AccountService for registration and login. Updated BreastCancerDB to use custom ApplicationRole, revised role seeding, and updated migrations and appsettings.json for JWT configuration. Removed all Keycloak-related code.

AmmarAbdelateif0-0 committed:
Merge branch 'main' into AddAccountController_V2

AmmarAbdelateif0-0 committed:
Refactor auth, add refresh tokens & patient CRUD APIs

- Implement refresh token support and robust JWT config
- Split registration by role; add role-specific DTOs
- Add PatientController with full CRUD and service layer
- Introduce AutoMapper and patient DTOs
- Add EmailService and controller (SMTP support)
- Update EF Core models, migrations for new relationships
- Improve error handling, logging, and validation
- Update dependencies and appsettings; remove Keycloak JSON

abdalahsalah commented on BreastCancer.csproj:
```diff
@@ -29,11 +30,7 @@
     <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.20.1" />
 	<PackageReference Include="Swashbuckle.AspNetCore" Version="6.5.0" />
 	<PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="8.15.0" />
-	<PackageReference Include="AutoMapper" Version="13.0.1" />
```
Why?

mahmoud-40 commented on Service/Implementation/AccountService.cs:
```diff
@@ -0,0 +1,277 @@
+
+        private readonly IAuthTokenService _authToken;
+
+        public AccountService(
```
You are not using the mapper

mahmoud-40 commented on Service/Implementation/AccountService.cs:
```diff
@@ -0,0 +1,277 @@
+                return (false, userResult.Errors);
+            }
+
+            var patient = new Patient
```
When I tried to use this endpoint it returned this error:
Error: response status is 500
`Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the entity changes. See the inner exception for details.
 ---> Microsoft.Data.SqlClient.SqlException (0x80131904): Cannot insert the value NULL into column 'DateOfBirth', table 'BreastCancerDB.dbo.AspNetUsers'; column does not allow nulls. INSERT fails.
The statement has been terminated.`

Same error happened with the rest of the register endpoints. 

I believe you should have tested them.

mahmoud-40 commented on Controllers/AccountController.cs:
```diff
@@ -0,0 +1,146 @@
+            this.accountService = accountService;
+        }
+        [HttpPost("Register")]
+        public async Task<IActionResult> Register(BaseRegisterDTO UserFromRequest)
```
This error happened when I tried to test this endpoint:

`Error: response status is 500

Response body
Download
System.NotSupportedException: Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported. Type 'BreastCancer.DTO.request.BaseRegisterDTO'. Path: $ | LineNumber: 0 | BytePositionInLine: 1.
 ---> System.NotSupportedException: Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported. Type 'BreastCancer.DTO.request.BaseRegisterDTO'.
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(ReadStack& state, Utf8JsonReader& reader, NotSupportedException ex)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.ContinueDeserialize(ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.DeserializeAsync(Stream utf8Json, CancellationToken cancellationToken)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.DeserializeAsObjectAsync(Stream utf8Json, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Formatters.SystemTextJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.Formatters.SystemTextJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value, Object container)`

I believe it's related to the abstract base register dto.

mahmoud-40 commented on Mapping/MappingProfile.cs:
```diff
@@ -9,8 +9,8 @@ public class MappingProfile : Profile
             CreateMap<Patient, PatientResponseDTO>()
-                .ForMember(dest => dest.Email, opt => opt.MapFrom(src => src.Email ?? string.Empty))
-                .ForMember(dest => dest.DoctorName, opt => opt.MapFrom(src => src.Doctor != null ? src.Doctor.FullName : null));
+                .ForMember(dest => dest.Email, opt => opt.MapFrom(src => src.User.Email ?? string.Empty))
```
Instead of mapping each member. 
You may use this:
` .ForAllMembers(opts => opts.Condition((src, dest, srcMember) => srcMember != null));`

mahmoud-40 commented on Service/Implementation/AccountService.cs:
```diff
@@ -0,0 +1,277 @@
+
+
+            // ToDo: AutoMapper
+            var user = new ApplicationUser
```
Please use AutoMapper.
It's a TODO.

mahmoud-40 commented on Controllers/AccountController.cs:
```diff
@@ -0,0 +1,146 @@
+
+            foreach (var error in result.Errors)
+            {
+                ModelState.AddModelError("", error);
```
When I attempted to submit invalid passwords like this input: 
Password: 123 
ConfirmPassword: 123 
I received the following error with status code 400: 
`{   "": [     "Passwords must be at least 8 characters long.",     "Passwords must contain at least one non-alphanumeric character.",     "Passwords must include at least one lowercase letter ('a'-'z').",     "Passwords must have at least one uppercase letter ('A'-'Z')."   ] }` 
I believe that in this case, it would be better to return a single clear message like: 
`{ "Password does not meet the required criteria." }` 
Or to provide individual error messages for each issue (which is recommended): 
`Passwords must be at least 8 characters long.` 
`Passwords must contain at least one non-alphanumeric character.` 
and so on.

mahmoud-40 reviewed (changes requested):
The endpoints did not function correctly during my testing. Please retest them, address any issues, and resolve all comments. Thank you.

mahmoud-40 commented:
<img width="1398" height="890" alt="image" src="https://github.com/user-attachments/assets/4a0ffb30-51d8-4e5c-8133-95b311c4d95d" />
<img width="1402" height="876" alt="image" src="https://github.com/user-attachments/assets/42a4ba69-26a2-4828-8c6b-fef87df85c2b" />

mahmoud-40 commented on DTO/request/BaseRegisterDTO.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace BreastCancer.DTO.request
+{
+    public abstract class BaseRegisterDTO
```
Also please review all the register dtos and the user (and each role) entity and make sure to add all the required fields (eg: Date of Birth)

AmmarAbdelateif0-0 committed:
merge with main

AmmarAbdelateif0-0 committed:
Refactor registration logic and enhance user mapping

- Remove generic Register endpoint; add specific Doctor, Patient, and Caregiver registration endpoints with correct success messages.
- Make Role property virtual in BaseRegisterDTO and override in derived DTOs for proper role assignment.
- Add DateOfBirth to BaseRegisterDTO and map to ApplicationUser.
- Add and register AutoMapper profiles for DTO/entity mapping.
- Restrict PatientController to "Patient" role only.
- Mark UserId as [Key] in Caregiver model.
- Refactor AccountService to use AutoMapper and remove generic RegisterAsync.
- Reduce JWT expiration time from 5 to 2 minutes.
- Add migration V13__ReDesgin for updated schema.
- Register IPatientService and PatientService in DI container.

AmmarAbdelateif0-0 committed:
Update refresh token time logic and JWT expiration config

Changed refresh token validation to use local time instead of UTC.
Increased JWT expiration time in appsettings.json from 2 to 20 minutes.
